File: crude-oil-forecast/commodity analysis.R
# Install and load necessary packages
install.packages(c("tseries", "forecast", "urca", "astsa"))
library(tseries)
library(forecast)
library(urca)
library(astsa)

# Load data
data_path <- "Modified Data.csv"  # Replace with your file path
data <- read.csv(data_path)
str(data)

# Convert to time series
price_ts <- ts(data = data$Price, start = c(2002, 1), frequency = 12)
str(price_ts)

# Plot time series
plot(price_ts)
autoplot(price_ts)

# Decompose time series
plot(decompose(price_ts))
seasonplot(price_ts, col = 1:20, pch = 19)

# Test for normality
shapiro.test(price_ts)

# Test for stationarity
ndiffs(price_ts)
nsdiffs(price_ts)

hist(price_ts)
acf(price_ts)
pacf(price_ts)
adf_price <- adf.test(price_ts)
print(adf_price)

# Differencing data
differenced_price <- diff(price_ts)
adf_differenced <- adf.test(differenced_price)

# Autocorrelation of differenced data
acf2(differenced_price)
acf(differenced_price, lag = 300)

# Fit ARIMA model
model <- auto.arima(differenced_price, trace = T)
arima_model <- arima(differenced_price, order = c(1, 1, 0))
residuals <- residuals(model)
plot(residuals)
checkresiduals(model)

# Forecast
forecasted <- forecast(arima_model, h = 24)
plot(forecasted)
sarima.for(differenced_price, 10, 1, 1)

# Check accuracy
accuracy(forecasted)

# Plot histogram
hist(price_ts, xlab = "Price", freq = F, col = "Blue", main = "Histogram of Commodity Price")
lines(density(price_ts), col = "green", lwd = 7)

# Calculate slope and intercept
price_times <- as.numeric(time(price_ts))
price_values <- as.numeric(price_ts)
SSxx <- sum((price_times - mean(price_times)) * (price_times - mean(price_times)))
SSxy <- sum((price_values - mean(price_values)) * (price_times - mean(price_times)))
slope <- SSxy / SSxx
intercept <- mean(price_values) - slope * mean(price_times)

File: crude-oil-forecast/commodity analysis.py
#%% 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from statsmodels.tsa.seasonal import seasonal_decompose
from statsmodels.graphics.tsaplots import plot_acf, plot_pacf
from statsmodels.tsa.stattools import adfuller
from pmdarima.arima import auto_arima, ndiffs, nsdiffs
from statsmodels.tsa.arima_model import ARIMA
from scipy.stats import shapiro

#%% Read data
data = pd.read_csv('Modified Data.csv')  # replace 'your_file.csv' with your file path
data_ts = data['Price']  # replace 'Price' with your column name

#%% Plot data
data_ts.plot()
plt.show()

#%% Decompose data
result = seasonal_decompose(data_ts, model='additive', period=12)
result.plot()
plt.show()

#%% Shapiro-Wilk test
print(shapiro(data_ts))

#%% Differencing
n_diffs = ndiffs(data_ts, test='adf')
print(f"\nNumber of differences required: {n_diffs}")

nsdiff= nsdiffs(data_ts, m=12, test='ch')
print(f"\nSeasonal differences required: {nsdiff}")

# %% Perform differencing if required
if n_diffs > 0:
    differenced_data = data_ts.diff(n_diffs).dropna()
else:
    differenced_data = data_ts.copy()


#%% ACF and PACF
plot_acf(data_ts)
plot_pacf(data_ts)
plt.show()

#%% ADF test
result = adfuller(data_ts)
print(f'ADF Statistic: {result[0]}')
print(f'p-value: {result[1]}')

#%% ARIMA model
model = auto_arima(differenced_data, trace=True)
model.fit(differenced_data)

#%% Forecast
forecast = model.forecast(n_periods=24)
plt.plot(forecast)
plt.show()

# Accuracy (replace 'test' with your test data)
# from sklearn.metrics import mean_squared_error
# print(np.sqrt(mean_squared_error(test, forecast)))

#%% Histogram
plt.hist(data_ts, color='blue')
plt.show()

#%% Density plot
data_ts.plot(kind='kde', color='green')
plt.show()

#%% Linear regression (replace 'time' with your time data)
slope, intercept = np.polyfit(data['time'], data_ts, 1)  # replace 'time' with your time column
print(f'Slope: {slope}, Intercept: {intercept}')

# %%
